libsql/internal/http/hranaV2/hranaV2.go

```diff
@@ -286,7 +287,119 @@ func sendPipelineRequest(ctx context.Context, msg *hrana.PipelineRequest, url st
 	return result, false, nil
 }
 
+func (h *hranaV2Conn) executeMsg(ctx context.Context, query string, msg *hrana.PipelineRequest) (*hrana.PipelineResponse, error) {
+	result, err := h.sendPipelineRequest(ctx, msg, false)
+	if err != nil {
+		return nil, fmt.Errorf("failed to execute SQL: %s\n%w", query, err)
+	}
+
+	for _, r := range result.Results {
+		if r.Error != nil {
+			return nil, fmt.Errorf("failed to execute SQL: %s\n%s", query, r.Error.Message)
+		}
+		if r.Response == nil {
+			return nil, fmt.Errorf("failed to execute SQL: %s\n%s", query, "no response received")
+		}
+	}
+	return result, nil
+}
+
+func (h *hranaV2Conn) executeInChunks(ctx context.Context, query string, wantRows bool) (*hrana.PipelineResponse, error) {
+	iterator := sqliteparserutils.CreateStatementIterator(query)
+
+	stmt, _, isEOF := iterator.Next()
+	// Skip all empty statements
+	for {
+		if stmt != "" {
+			break
+		}
+		stmt, _, isEOF = iterator.Next()
+		if isEOF {
+			break
+		}
+	}
+	if isEOF {
+		msg := &hrana.PipelineRequest{}
+		executeStream, err := hrana.ExecuteStream(stmt, nil, wantRows)
+		if err != nil {
+			return nil, fmt.Errorf("failed to execute SQL: %s\n%w", query, err)
+		}
+		msg.Add(*executeStream)
+		return h.executeMsg(ctx, query, msg)
+	}
+
+	sqls := make([]string, 4096)
+
+	batch := &hrana.Batch{Steps: make([]hrana.BatchStep, len(sqls))}
+	msg := &hrana.PipelineRequest{}
+	msg.Add(hrana.StreamRequest{Type: "batch", Batch: batch})
+	for idx := range batch.Steps {
+		batch.Steps[idx].Stmt.WantRows = wantRows
+	}
+
+	result := &hrana.PipelineResponse{}
+
+	sqls[0] = "BEGIN"
+	batch.Steps[0].Stmt.Sql = &sqls[0]
+	nextIdx := 1
+	first := true
+	for {
+		if stmt != "" {
+			sqls[nextIdx] = stmt
+			batch.Steps[nextIdx].Stmt.Sql = &sqls[nextIdx]
+			nextIdx += 1
+			if nextIdx == len(sqls) {
+				res, err := h.executeMsg(ctx, query, msg)
+				if err != nil {
+					h.closeStream()
+					return nil, err
+				}
+				result.Baton = res.Baton
+				result.BaseUrl = res.BaseUrl
+				idx := 0
+				if first {
+					idx = 1
+				}
+				for ; idx < len(res.Results); idx++ {
+					result.Results = append(result.Results, res.Results[idx])
+				}
+				first = false
+				nextIdx = 0
+			}
+		}
+		if isEOF {
+			break
+		}
+		stmt, _, isEOF = iterator.Next()
+	}
+	sqls[nextIdx] = "COMMIT"
```

Comment:
Do we need to close stream after commit?

Reply:
We don't have to and we probably also shouldn't.